### ISSUE TITLE
Return block height info along nullifiers and transaction ids

### ISSUE THREAD

**Issue by mFragaBA** [0xPolygonMiden/miden-node#380]
>> One thing I noticed we might still have problems with in the client is that we're still not getting the info of _when_ the transaction was included in a block. This is due to the fact that we return a sync state response with all nullifiers and transaction ids in the range (request.block_header, next_block_header) where block header is the block of the first note matching any of the requests' tags or account ids. So, if many TXs that only consume notes are sent and then we sync after a while, we'll still receive the TXs ids but won't know which got included in which block. A similar thing happens with nullifiers as well.
>
>Ah - good point! We could return `(TransactionId, block_num)` and `(Nullifier, block_num)` tuples. But that's separate from this PR. Could you create a new issue for this?

_Originally posted by @bobbinth in https://github.com/0xPolygonMiden/miden-node/issues/377#issuecomment-2155540500_

**Comment by bobbinth** [comment]
As a part of this, let's do the following:

1. Investigate moving transaction data into `AccountSummary` (rather than having it as a separate filed in the response).
2. Update endpoint docs in readme files (i.e., [here](https://github.com/0xPolygonMiden/miden-node/tree/next/crates/rpc#syncstate) and [here](https://github.com/0xPolygonMiden/miden-node/blob/next/crates/store/README.md#syncstate)).

**Comment by mFragaBA** [comment]
small update:

- I checked and the node already sends the block height for each `Nullifier`, we just were not using the info on the client side
- as for the transactions, I made a test branch named `﻿mFragaBA-return-tx-and-nullifier-block-height-info` with the "naive" implementation (just adding the block height along with the ID)
- I'm gonna follow up with another branch where I change the AccountSummary message as said [here](https://github.com/0xPolygonMiden/miden-node/issues/371#issuecomment-2143112925). I think it's doable, though I still don't quite understand the use case (my thought process is that a client will only care about the txs generated by them, so they should have all the info needed).

- I still need to update the docs, but I'll do that once we decide for either option
- I'll also update the client's code once we decide on either option

**Comment by igamigo** [comment]
When you mention "either option", do you mean coupling the block height to the transaction ID versus updating `AccountSummary`?
I disagree that the client would only care about transactions created locally. I think I mentioned it in the issue or we discussed offline, but for public accounts it makes sense that you can understand if and how many transactions were executed. In any case it should be marginal impact for a good amount of usable information: once you are tracking more than one account you wouldn't be able to tell how many transactions were executed on each account otherwise.

**Comment by bobbinth** [comment]
I agree with @igamigo - I think linking transactions to accounts is a useful property at relatively small extra cost.

I do wonder if trying send this info in a non-naive way (i.e., by changing `AccountSummary`) is a bit more complex than we originally thought, and if so, whether it is worth it vs. the naive way. Specifically, we'd need to change protobuf messages to something like this:

```proto
message AccountTransactionInfo {
    digest.Digest transaction_id = 1;
    fixed32 block_num = 2;
}

message AccountSummary {
    // ID of an account which had at least one transaction executed against it during the sync timeframe.
    AccountId account_id = 1;
    // A list of (transaction_id, block_num) tuples for transactions executed against the account during
    // the sync timeframe. This list is guaranteed to have at least one entry.
    repeated AccountTransactionInfo transactions = 2;
    // Commitment to the final state of the account updated during the sync response timeframe. If the
    // account state was not update during this timeframe, account_hash will be omitted from the response.
    optional digest.Digest account_hash = 3;
}
```

There are two potential issues with this approach:

1. If we update `AccountSummary` this way, we can no longer use it in the `AccountInfo` message because this is used in other places. Not sure how intrusive of a change it would be to decouple them (maybe nothing too bad).
2. I wonder if the meaning of `AccountSummary.account_hash` is somewhat confusing here. Basically, we can set it only if the account was updated during the sync timeframe, but not after. So, the following situation may happen:

The client is currently at block 1000. The account it monitors was updated at block 1100 and 1200. The client starts making sync request and the first request syncs the client to block 1150. Since there was a transaction against the account at block 1100, we'd get the account summary for this account, but `account_hash` would be empty since there was another update that happened after 1150 and so hash will be returned when we sync beyond block 1200.

I'm not sure how big of a problem this is, and this issue still kind of exists with the naive approach - so, maybe I'm overthinking it.

**Comment by mFragaBA** [comment]
Yesterday I did a somewhat similar approach to this in `mFragaBA-return-tx-info-on-account-summary`, and came up with similar issues:

> 1. If we update AccountSummary this way, we can no longer use it in the AccountInfo message because this is used in other places. Not sure how intrusive of a change it would be to decouple them (maybe nothing too bad).

This indeed happened and so instead of modifying `AccountSummary` I wrapped it with a `AccountUpdate` (I'm not so keen on that naming though)

> 2. I wonder if the meaning of AccountSummary.account_hash is somewhat confusing here. Basically, we can set it only if the account was updated during the sync timeframe, but not after. So, the following situation may happen:

Agreed, since we don't keep in the store a history of account summaries it's not as easy as building the `AccountSummary` using the account corresponding to the transaction's account_id because we might have already lost the account state in that time frame.

I think using the naive approach and sending the transactions as a separate field is a better approach, and we can include the transactions' account ids as well. For example, we could define:

```proto
message TransactionUpdate {
    digest.Digest transaction_id = 1;
    fixed32 block_num = 2;
    account.AccountId account_id = 3;
}
```

and the `SyncStateResponse` would be:

```proto
message SyncStateResponse {
    // ...

    // List of IDs for transactions executed against requested accounts between
    // `block_num + 1` and `block_header.block_num`
    repeated TransactionUpdate transactions = 6;

    // ...
}
```

If we wanted we can define it as a map to avoid some account id redundance

---

#### Some questions I came up with since I'm not sure we have a standardized way of defining protobuf messages

> ```proto
> message AccountTransactionInfo {
>     digest.Digest transaction_id = 1;
>     fixed32 block_num = 2;
> }
> ```

I noticed some messages such as  `AccountSummary` use `uint32` instead of `fixed32` for the block num. Is there any difference between using either?

Another thing I noticed is that for `AccountId` we have a protobuf message

```proto
message AccountId {
    // A miden account is defined with a little bit of proof-of-work, the id itself is defined as
    // the first word of a hash digest. For this reason account ids can be considered as random
    // values, because of that the encoding bellow uses fixed 64 bits, instead of zig-zag encoding.
    fixed64 id = 1;
}

message AccountSummary {
    AccountId account_id = 1;
    digest.Digest account_hash = 2;
    uint32 block_num = 3;
}
```

Should I also build a message for `TransactionId` instead of using `digest.Digest`?

**Comment by mFragaBA** [comment]
I pushed some changes to `mFragaBA-return-tx-and-nullifier-block-height-info` (the decoupled version) to also include the account IDs in the response. You can see the changes [here](https://github.com/0xPolygonMiden/miden-node/compare/mFragaBA-return-tx-and-nullifier-block-height-info?expand=1). If you're ok with the decoupled approach I'll update the related documentation and open a PR

**Comment by bobbinth** [comment]
> I think using the naive approach and sending the transactions as a separate field is a better approach

I'm leaning this way as well. The info we send is identical. One can argue that putting transaction data into `AccountSummary` is a bit more compact and will make it a bit easier to process it on the client - but I think the benefit will be marginal (if any).

> I noticed some messages such as `AccountSummary` use `uint32` instead of `fixed32` for the block num. Is there any difference between using either?

`uint32` will be encoded as a variable-length integer while `fixed32` is always encoded as 4 bytes (I believe). So, for large values (e.g., block numbers) it makes sense to use `fixed32`.

> Should I also build a message for `TransactionId` instead of using `digest.Digest`?

Yes, I think that's a good idea.

**Comment by bobbinth** [comment]
Closed by #386.
